Implement Bulk TSSC Transfer in DTP CLI via Fund contract

abhi3700 opened the pull request:
## Description
Implemented a new `Fund` contract to facilitate bulk transfers of TSSC tokens with a single signature from the funder, improving upon the traditional method of signing each transfer individually using the `eth_sendRawTransaction` chain API call in `transfer_tssc_single`.

This update enables all transfers to be added in a single transaction within a block, as opposed to the previous iterative approach used in the `transfer_tssc_single` function, which resulted in multiple transactions being added to different blocks with a gap of 1-5 blocks between each transfer.

## Key changes
- [Introduced a Fund contract](https://github.com/abhi3700/subspace-evm-playground/compare/d1da2720fa4576103414e830a56b356b746625b7...d6346ddc199cfc66f803bbd76ac9e54152fb86db#diff-3e6016c675e31e69d5a1a6a624735fd8142a34282dbb47b9b39422aa619a79cb) for bulk TSSC token transfers with a single funder signature, replacing the previous iterative `eth_sendRawTransaction` approach.
- Debugged [`transferTsscToMany`](https://github.com/abhi3700/subspace-evm-playground/compare/d1da2720fa4576103414e830a56b356b746625b7...d6346ddc199cfc66f803bbd76ac9e54152fb86db#diff-3e6016c675e31e69d5a1a6a624735fd8142a34282dbb47b9b39422aa619a79cbR36-R48) for multiple receivers, added corresponding contract [tests](https://github.com/abhi3700/subspace-evm-playground/compare/d1da2720fa4576103414e830a56b356b746625b7...d6346ddc199cfc66f803bbd76ac9e54152fb86db#diff-63357471d2fb30b4517087c3ec467a3587bb761eacb505c1962b627788a90343), [deploy script](https://github.com/abhi3700/subspace-evm-playground/compare/d1da2720fa4576103414e830a56b356b746625b7...d6346ddc199cfc66f803bbd76ac9e54152fb86db#diff-dd3e0aa5776d459b525ce552f3aa69b088e3113c6925f8b9efa5fec5921f2786) and modified the function accordingly.
- Kept [`transferTsscToSingle`](https://github.com/abhi3700/subspace-evm-playground) as a private method and called within [`transferTsscToMany`](https://github.com/abhi3700/subspace-evm-playground/compare/d1da2720fa4576103414e830a56b356b746625b7...d6346ddc199cfc66f803bbd76ac9e54152fb86db#diff-3e6016c675e31e69d5a1a6a624735fd8142a34282dbb47b9b39422aa619a79cbR36-R48).
- [Renamed `transfer_tssc` to `transfer_tssc_single`](https://github.com/abhi3700/domain-transaction-producer/pull/2/files#diff-c730d471704d7db227e676d88503447de9b5463c8206b949b66dbb6da2211073R373) in **dtp** package of the workspace.
- Added `transfer_tssc_bulk` function in **dtp** package for single-transaction bulk transfers, and archived the [now-obsolete `transfer_tssc_single` function](https://github.com/abhi3700/domain-transaction-producer/pull/2/files#diff-c730d471704d7db227e676d88503447de9b5463c8206b949b66dbb6da2211073R342-R346).
- [Added `FUND` contract address into `.env.example` file](https://github.com/abhi3700/domain-transaction-producer/pull/2/commits/5723c8319aed906239474b2dcdca8645d6e01066).

abhi3700 committed:
Implement Bulk Transfer in DTP CLI via Fund contract

- Introduced a Fund contract for bulk TSSC token transfers with a single funder signature, replacing the previous iterative `eth_sendRawTransaction` approach.
- Debugged `transferTsscToMany` for multiple receivers, added corresponding contract tests, and modified the function accordingly.
- Kept `transferTsscToSingle` as a private method and called within `transferTsscToMany`.
- Added transfer_tssc_bulk in DTP CLI for single-transaction bulk transfers, and archived the now-obsolete `transfer_tssc_single` function.

abhi3700 committed:
added the deployed Fund contract address in '.env.example' file

0xjeremyfrank commented on dtp/src/utils.rs:
```diff
@@ -404,15 +339,153 @@ pub(crate) async fn gen_wallets_transfer_tssc(
+        // FIXME: Failure in getting pending tx: Revert(Bytes(0x))
+        .expect("Failure in getting pending tx")
+        .await?
+        .expect("Failure in \'transferTsscToMany\' of Fund contract");
```
```suggestion
        .expect("Failure in \'transferTsscTooMany\' of Fund contract");
```

abhi3700 replied:
Here, are you suggesting to change the function name of the contract or just as error message?
Actually, I referred to `transferTsscToMany` as the contract method being called here.

0xjeremyfrank replied:
nevermind, I think i misread this. ignore. 🤦‍♂️

0xjeremyfrank commented on dtp/src/utils.rs:
```diff
@@ -404,15 +339,153 @@ pub(crate) async fn gen_wallets_transfer_tssc(
+/// the contract "Fund"'s `transferTsscToMany`
+/// One can also leverage the contract "Fund"'s `transferTsscToSingle`
+/// function.
+pub(crate) async fn transfer_tssc_single(
```
this can probably be removed.

0xjeremyfrank replied:
I mean remove the function completely, I don't think we will use it.

abhi3700 replied:
Done in this [commit](https://github.com/abhi3700/domain-transaction-producer/pull/2/commits/5e4843e117e797fac9a346966289fdcdab91f9af#).

0xjeremyfrank reviewed (commented):
The bulk transfer works great! I was able to fund > 1,000 accounts before hitting gas limit. I had a couple of comments that should be addressed.

However, sending subsequent transactions doesn't work at scale, it breaks when passed 1000 accounts with a "Too many connections" error. I'm not sure the underlying cause, but much of the code seems inefficient in the # of calls made. These inefficiencies are outside of the code in this PR but include calling  `get_chainid` in every iteration, and in `get_gas_cost` multiple calls to `get_transaction`. I didn't do an exhaustive look, but I think for the next PR you should try to make the code much more efficient and test at a bit higher scale locally.

abhi3700 commented:
Noted!

For the error: "Too many connections", I could use batching with small chunks of 100 requests. I ran some testing with different chunk value - 100, 200, 300, 400. These are my observations:

- `100`: 100% success.
- `200`: 70% success, not sure.
- `300`:  50% success, not sure.
- `400`: 100% failure. 

I will take this up in the next PR #3. For now, following your code changes & cleaning for this PR.

abhi3700 committed:
removed 'transfer_tssc_single' function

Not needed anymore as there is a better function for bulk transfer: 'transfer_tssc_bulk'

abhi3700 committed:
allowing some function to be used later

abhi3700 commented:
Addressed to the concerns in PR #3.
Ready to close this PR.
